### PR TITLE
graceful shutdown fix

### PR DESCRIPTION
**Pull request by GrigoriyMikhalkin**
There's few problems with a graceful shutdown that i tried to fix with this PR.

1. `shutdownChan` channel that was used to finish `PollNetMapStream` didn't work if no `PollNetMapStream` was running at the time of shutdown. Also, it could stop only single `PollNetMapStream` routine, as single value was passed to channel instead of closing it.
2. There is a race condition with function that handles termination signals, as it isn't included in an error group.

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -96,6 +96,7 @@ type Headscale struct {
 	ipAllocationMutex sync.Mutex
 
 	shutdownChan chan struct{}
+	wg           sync.WaitGroup
```
Comment (kradalby) [inline app.go:99]: Could we give this a more descriptive name? what is the group waiting for?

File: app.go [hunk 2]
```diff
@@ -677,7 +681,7 @@ func (h *Headscale) Serve() error {
 		syscall.SIGTERM,
 		syscall.SIGQUIT,
 		syscall.SIGHUP)
-	go func(c chan os.Signal) {
+	sig_func := func(c chan os.Signal) {
```
Comment (kradalby) [inline app.go:684]: sigFunc

**Review by kradalby — COMMENTED** [review]
Looks reasonable, just some small nits

**Comment by GrigoriyMikhalkin** [comment]
@kradalby I additionally decomposed `OIDCCallback` method. Otherwise `gocyclo` complaint about it's complexity.

File: app.go (hunk 1 above)
Comment (GrigoriyMikhalkin) [reply app.go]: Updated the name. This wait group is for `PollNetMap` streams.

**Comment by kradalby** [comment]
> @kradalby I additionally decomposed `OIDCCallback` method. Otherwise `gocyclo` complaint about it's complexity. 

Can you separate this into a different PR?

**Comment by GrigoriyMikhalkin** [comment]
> > @kradalby I additionally decomposed `OIDCCallback` method. Otherwise `gocyclo` complaint about it's complexity.
> 
> Can you separate this into a different PR?

Done

**Review by juanfont — COMMENTED** [review]
Only one minor thing to be fixed.

File: app.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (GrigoriyMikhalkin) [reply app.go]: That's already fixed in recent commits) https://github.com/juanfont/headscale/pull/668/files#diff-6c4b6ed7dc8834cef100f50dae61c30ffe7775a3f3f6f5a557517cb740c44a2dR675

**Review by juanfont — APPROVED** [review]

**Comment by juanfont** [comment]
LGTM!

**Pull request merged**